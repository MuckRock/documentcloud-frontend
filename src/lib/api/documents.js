/** API helpers related to documents.
 * Lots of duplicated code here that should get consolidated at some point.
 */
import { error } from "@sveltejs/kit";
import { BASE_API_URL } from "@/config/config.js";
import { DEFAULT_EXPAND } from "@/api/common.js";

/**
 * Search documents
 *
 * @async
 * @param {query} string
 * @param {boolean} highlight
 * @param {globalThis.fetch} fetch
 * @returns {import('./types').DocumentResults}
 */
export async function search(
  query = "",
  highlight = false,
  fetch = globalThis.fetch,
) {
  const endpoint = new URL("documents/search/", BASE_API_URL);

  endpoint.searchParams.set("expand", DEFAULT_EXPAND);
  endpoint.searchParams.set("q", query);
  endpoint.searchParams.set("hl", highlight);

  const resp = await fetch(endpoint, { credentials: "include" });

  if (!resp.ok) {
    error(resp.status, resp.statusText);
  }

  return resp.json();
}

/**
 * Load a single document from the API
 * Example: https://api.www.documentcloud.org/api/documents/1/
 *
 * @async
 * @export
 * @param {number} id
 * @param {globalThis.fetch} fetch
 * @returns {import('./types').Document}
 */
export async function get(id, fetch) {
  const endpoint = new URL(`documents/${id}.json`, BASE_API_URL);
  const expand = ["user", "organization", "projects", "revisions"];
  endpoint.searchParams.set("expand", expand.join(","));

  const resp = await fetch(endpoint, { credentials: "include" });

  if (!resp.ok) {
    error(resp.status, resp.statusText);
  }

  return resp.json();
}

// utility functions

/**
 * Canonical URL for a document, relative to the current server
 * This will be correct in all environments, including deploy previews
 *
 * @export
 * @param {import('./types').Document} document
 * @returns {URL}
 */
export function canonicalUrl(document) {}
/**
 * Canonical URL for a single page embed, relative to the current server
 * This will be correct in all environments, including deploy previews
 *
 * @export
 * @param {import('./types').Document} document
 * @param {number} page
 * @returns {URL}
 */
export function canonicalPageUrl(document, page) {}

/**
 * Hash URL for a single page within the document viewer
 *
 * @export
 * @param {import('./types').Document} document
 * @param {number} page
 * @returns {URL}
 */
export function pageUrl(document, page) {}

/**
 * Static URL for a page image
 *
 * @export
 * @param {import('./types').Document} document
 * @param {number} page
 * @param {import('./types').sizes} size
 * @returns {URL}
 */
export function pageImageUrl(document, page, size) {}

/**
 * Asset URL for page text
 *
 * @export
 * @param {import('./types').Document} document
 * @param {number} page
 * @returns {URL}
 */
export function textUrl(document, page) {}
