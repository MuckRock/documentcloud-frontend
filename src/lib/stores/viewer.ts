import { writable, type Writable } from "svelte/store";
import type { Note, ViewerMode } from "../api/types";

export const currentPage: Writable<number> = writable(1);
export const activeNote: Writable<Note> = writable(null);
export const currentMode: Writable<ViewerMode> = writable("document");
