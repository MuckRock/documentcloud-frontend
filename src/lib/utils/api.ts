import type { NumericRange } from "@sveltejs/kit";
import type { APIResponse, Maybe, Page } from "$lib/api/types";
import { CSRF_COOKIE_NAME, MAX_PER_PAGE } from "@/config/config";

export function isErrorCode(status: number): status is NumericRange<400, 599> {
  return status >= 400 && status <= 599;
}

export function isRedirectCode(
  status: number,
): status is NumericRange<300, 308> {
  return status >= 300 && status <= 308;
}

/**
 * Handle what comes back from the API and return either data or errors.
 *
 * Two generic types are passed through:
 *
 * - T is data from the API
 * - E is an error coming back from the API
 *
 * @param resp The fetch response from the API. If this is missing, fetch
 * threw an error and we should send a 500 to the user because the API is
 * probably down.
 */
export async function getApiResponse<T, E = unknown>(
  resp?: Response | void,
): Promise<APIResponse<T, E>> {
  const response: APIResponse<T, E> = {};

  if (!resp) {
    response.error = {
      status: 500,
      message: "API error",
    };

    return response;
  }

  if (isErrorCode(resp.status)) {
    try {
      response.error = {
        status: resp.status,
        message: resp.statusText,
        errors: resp.json ? ((await resp.json()) as E) : undefined,
      };
    } catch (error) {
      console.warn(error);
      // if we fail parsing the error's JSON,
      // just return the status
      response.error = {
        status: resp.status,
        message: resp.statusText,
        errors: undefined,
      };
    }

    return response;
  }

  // everything worked

  if (resp.status === 204) {
    // deletes return nothing
    return {};
  }

  try {
    // redactions return an empty 200 response
    response.data = resp.json ? ((await resp.json()) as T) : ({} as T);
  } catch (e) {
    switch (e.name) {
      case "SyntaxError": // provide more specific error
        response.error = {
          status: 500,
          message: "The API returned invalid JSON",
        };
        break;
      default: // catch-all handling
        response.error = {
          status: 500,
          message: String(e),
        };
    }
  }
  return response;
}

/**
 * Requests the specified URL and paginates through to return all
 * results if additional pages are present.
 */
export async function getAll<T>(
  url: URL,
  perPage: number = MAX_PER_PAGE,
  fetch = globalThis.fetch,
): Promise<Array<T>> {
  url.searchParams.set("per_page", String(perPage));
  const resp = await fetch(url, { credentials: "include" });
  const data: Page<T> = await resp.json();
  const results = data.results;
  if (data.next) {
    // Recursively fetch the next page
    const nextResults = await getAll<T>(new URL(data.next), perPage, fetch);
    return results.concat(nextResults);
  }
  return results;
}

/**
 * Handle the two-step redirect needed to get a private asset from S3.
 */
export async function getPrivateAsset(
  url: URL | string,
  fetch = globalThis.fetch,
): Promise<URL> {
  // any errors here should be caught higher up
  const resp = await fetch(url, {
    credentials: "include",
    redirect: "error",
    headers: {
      Accept: "application/json",
    },
  }).catch(console.warn);

  // should this be an API response?
  if (!resp) {
    throw new Error("API error");
  }

  if (!resp.ok) {
    throw new Error(resp.statusText);
  }

  // {"location": "s3 URL with credentials"}
  const { location } = (await resp.json()) as { location: string };

  // if this isn't a valid URL, it's an error
  return new URL(location);
}

export function getCsrfToken(document = globalThis.document): Maybe<string> {
  if (typeof document === "undefined") return "";
  const [key, token] =
    document.cookie
      ?.split(";")
      ?.map((c) => c.split("="))
      // in case there's spaces in the cookie string, trim the key
      ?.find(([k, v]) => k?.trim() === CSRF_COOKIE_NAME) ?? [];

  return token;
}
