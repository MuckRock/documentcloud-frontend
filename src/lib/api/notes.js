import { error } from "@sveltejs/kit";
import { BASE_API_URL } from "@/config/config.js";
import { DEFAULT_EXPAND } from "@/api/common.js";
import { canonicalUrl } from "./documents";

/**
 * Load notes from a single document from the API
 * Example: https://api.www.documentcloud.org/api/documents/2622/notes/
 *
 * @async
 * @export
 * @param {number} doc_id
 * @param {globalThis.fetch} fetch
 * @returns {import('./types').NoteResults}
 */
export async function list(doc_id, fetch) {
  const endpoint = new URL(`documents/${doc_id}/notes.json`, BASE_API_URL);

  endpoint.searchParams.set("expand", DEFAULT_EXPAND);

  const resp = await fetch(endpoint, { credentials: "include" });

  if (!resp.ok) {
    error(resp.status, resp.statusText);
  }

  return resp.json();
}

/**
 * Load a single note from a single document from the API
 * Example: https://api.www.documentcloud.org/api/documents/2622/notes/549/
 *
 * @async
 * @export
 * @param {number} doc_id
 * @param {globalThis.fetch} fetch
 * @returns {import('./types').Note}
 */
export async function get(doc_id, note_id, fetch) {
  const endpoint = new URL(
    `documents/${doc_id}/notes/${note_id}.json`,
    BASE_API_URL,
  );

  endpoint.searchParams.set("expand", DEFAULT_EXPAND);

  const resp = await fetch(endpoint, { credentials: "include" });

  if (!resp.ok) {
    error(resp.status, resp.statusText);
  }

  return resp.json();
}

/**
 * Canonical URL for a note, relative to the current server
 * This will be correct in all environments, including deploy previews
 * https://www.documentcloud.org/documents/2622-agreement-between-conservatives-and-liberal-democrats-to-form-a-coalition-government/annotations/557
 *
 * @export
 * @param {import('./types').Document} document
 * @param {import('./types').Note} note
 * @returns {URL}
 */
export function canonicalNoteUrl(document, note) {
  return new URL(`annotations/${note.id}/`, canonicalUrl(document));
}

/**
 * Hash URL for a note within the document viewer
 * https://www.documentcloud.org/documents/2622-agreement-between-conservatives-and-liberal-democrats-to-form-a-coalition-government/#document/p3/a557
 *
 * @export
 * @param {import('./types').Document} document
 * @param {import('./types').Note} note
 * @returns {URL}
 */
export function noteUrl(document, note) {
  return new URL(
    `#document/p${note.page_number + 1}/a${note.id}`,
    canonicalUrl(document),
  );
}

/**
 * Width of a note, relative to the document
 *
 * @export
 * @param {import('./types').Note} note
 * @returns {number}
 */
export function width(note) {
  return note.x2 - note.x1;
}

/**
 * Height of a note, relative to the document
 *
 * @export
 * @param {import('./types').Note} note
 * @returns {number}
 */
export function height(note) {
  return note.y2 - note.y1;
}
