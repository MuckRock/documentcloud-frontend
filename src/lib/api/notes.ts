import type {
  APIResponse,
  BBox,
  Document,
  Note,
  NoteResults,
  ValidationError,
} from "./types";

import {
  APP_URL,
  BASE_API_URL,
  CSRF_HEADER_NAME,
  EMBED_URL,
} from "@/config/config.js";
import { DEFAULT_EXPAND } from "@/api/common.js";
import { canonicalUrl } from "./documents";
import { getApiResponse, isErrorCode } from "../utils";

/**
 * Load notes from a single document from the API
 * @example https://api.www.documentcloud.org/api/documents/2622/notes/
 * @deprecated
 */
export async function list(doc_id: number, fetch = globalThis.fetch) {
  const endpoint = new URL(`documents/${doc_id}/notes/`, BASE_API_URL);

  endpoint.searchParams.set("expand", DEFAULT_EXPAND);

  const resp = await fetch(endpoint, { credentials: "include" });

  if (isErrorCode(resp.status)) {
    throw new Error(resp.statusText);
  }

  return resp.json();
}

/**
 * Load a single note from a single document from the API
 * @example https://api.www.documentcloud.org/api/documents/2622/notes/549/
 */
export async function get(
  doc_id: number,
  note_id: number,
  fetch = globalThis.fetch,
): Promise<APIResponse<Note, unknown>> {
  const endpoint = new URL(
    `documents/${doc_id}/notes/${note_id}/`,
    BASE_API_URL,
  );

  endpoint.searchParams.set("expand", DEFAULT_EXPAND);

  const resp = await fetch(endpoint, { credentials: "include" });

  return getApiResponse<Note>(resp);
}

// writing methods

/**
 * Create a new note
 */
export async function create(
  doc_id: string | number,
  note: Partial<Note>,
  csrf_token: string,
  fetch = globalThis.fetch,
): Promise<APIResponse<Note, ValidationError>> {
  const endpoint = new URL(`documents/${doc_id}/notes/`, BASE_API_URL);

  const resp = await fetch(endpoint, {
    body: JSON.stringify(note),
    credentials: "include",
    headers: {
      "Content-type": "application/json",
      [CSRF_HEADER_NAME]: csrf_token,
      Referer: APP_URL,
    },
    method: "POST",
  });

  return getApiResponse<Note, ValidationError>(resp);
}

/**
 * Update a note and return the updated version
 */
export async function update(
  doc_id: string | number,
  note_id: string | number,
  note: Partial<Note>,
  csrf_token: string,
  fetch = globalThis.fetch,
): Promise<APIResponse<Note, ValidationError>> {
  const endpoint = new URL(
    `documents/${doc_id}/notes/${note_id}/`,
    BASE_API_URL,
  );

  const resp = await fetch(endpoint, {
    body: JSON.stringify(note),
    credentials: "include",
    headers: {
      "Content-type": "application/json",
      [CSRF_HEADER_NAME]: csrf_token,
      Referer: APP_URL,
    },
    method: "PATCH",
  });

  return getApiResponse<Note, ValidationError>(resp);
}

/**
 * Delete a note
 */
export async function remove(
  doc_id: string | number,
  note_id: string | number,
  csrf_token: string,
  fetch = globalThis.fetch,
) {
  const endpoint = new URL(
    `documents/${doc_id}/notes/${note_id}/`,
    BASE_API_URL,
  );

  const resp = await fetch(endpoint, {
    credentials: "include",
    headers: {
      "Content-type": "application/json",
      [CSRF_HEADER_NAME]: csrf_token,
      Referer: APP_URL,
    },
    method: "DELETE",
  });

  return getApiResponse<null>(resp);
}

/**
 * Canonical URL for a note, relative to the current server
 * This will be correct in all environments, including deploy previews
 * @example https://www.documentcloud.org/documents/2622-agreement-between-conservatives-and-liberal-democrats-to-form-a-coalition-government/annotations/557
 */
export function canonicalNoteUrl(document: Document, note: Note): URL {
  return new URL(
    `/documents/${document.id}/annotations/${note.id}/`,
    EMBED_URL,
  );
}

/**
 * Hash URL for a note within the document viewer
 * @example https://www.documentcloud.org/documents/2622-agreement-between-conservatives-and-liberal-democrats-to-form-a-coalition-government/#document/p3/a557
 */
export function noteUrl(document: Document, note: Note): URL {
  return new URL(
    `#document/p${note.page_number + 1}/a${note.id}`,
    canonicalUrl(document),
  );
}

/**
 * Generate the hash URL for a note, without the document URL
 * @param note
 * @returns hash
 */
export function noteHashUrl(note: Note): string {
  return `#document/p${note.page_number + 1}/a${note.id}`;
}

/**
 * Opposite of noteHashUrl, returning a note ID.
 * To get the page number, use pageFromHash
 * @param hash
 */
export function noteFromHash(hash: string): number {
  const re = /^#document\/p(\d+)\/a(\d+)$/;
  const match = re.exec(hash);

  if (!match) return null;

  return +match[2] || null;
}

/** Width of a note, relative to the document */
export function width(note: BBox): number {
  return note.x2 - note.x1;
}

/** Height of a note, relative to the document */
export function height(note: BBox): number {
  return note.y2 - note.y1;
}

export function isPageLevel(note: Partial<Note>): boolean {
  if (!note) return false;
  return [note?.x1, note?.x2, note?.y1, note?.y2].every(
    (n) => n === undefined || n === null,
  );
}
