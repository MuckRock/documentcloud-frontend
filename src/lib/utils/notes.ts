import type { PDFDocumentProxy } from "pdfjs-dist/types/src/display/api";
import type { Access, Document, Note, Sizes } from "$lib/api/types";

import { pageImageUrl } from "$lib/api/documents";
import { width, height } from "$lib/api/notes";

const SIZE: Sizes = "large";
const PADDING = 50; // vertical padding in pixels

export function getCanvasContext(
  canvas: HTMLCanvasElement,
): CanvasRenderingContext2D | null {
  const context = canvas.getContext("2d");
  if (!context) {
    console.error("Missing canvas context when rendering note.");
  }
  return context;
}

interface ContextDimensions {
  sourceY: number;
  sourceHeight: number;
  noteWidth: number;
  noteHeight: number;
}

export function calculateImageContextDimensions(
  note: Note,
  image: HTMLImageElement,
): ContextDimensions {
  const noteWidth = width(note) * image.width;
  const noteHeight = height(note) * image.height;
  const sourceY = Math.max(0, note.y1 * image.height - PADDING);
  const sourceHeight = Math.min(
    image.height - sourceY,
    noteHeight + 2 * PADDING,
  );

  return { sourceY, sourceHeight, noteWidth, noteHeight };
}

export function calculatePDFContextDimensions(
  note: Note,
  pageWidth: number,
  pageHeight: number,
  scale: number,
): ContextDimensions {
  const noteWidth = width(note) * pageWidth * scale;
  const noteHeight = height(note) * pageHeight * scale;
  const sourceY = Math.max(0, note.y1 * pageHeight * scale - PADDING);
  const sourceHeight = Math.min(
    pageHeight * scale - sourceY,
    noteHeight + 2 * PADDING,
  );

  return { sourceY, sourceHeight, noteWidth, noteHeight };
}

export function drawHighlight(
  context: CanvasRenderingContext2D,
  fillStyle: string,
  x: number,
  y: number,
  width: number,
  height: number,
): void {
  context.save();
  context.fillStyle = fillStyle;
  context.globalAlpha = 0.5;
  context.globalCompositeOperation = "multiply";
  context.fillRect(x, y, width, height);
  context.restore();
}

export function setupCanvas(
  canvas: HTMLCanvasElement,
  width: number,
  height: number,
  dpr: number = 1,
): void {
  canvas.width = Math.floor(width * dpr);
  canvas.height = Math.floor(height * dpr);
}

function computeCssVar(varName: string, fallback: string): string {
  const computedStyle = getComputedStyle(globalThis.document.documentElement);
  return computedStyle.getPropertyValue(varName).trim() || fallback;
}

export const DEFAULT_COLORS: Record<Access, string> = {
  public: "#eccb6b",
  private: "#4294f0",
  organization: "#27c6a2",
} as const;

export const DEFAULT_FALLBACK_COLOR = "#d8dee2";

export function isServerSide(): boolean {
  return typeof window === "undefined";
}

export function getAccessColor(access: Access): string {
  return DEFAULT_COLORS[access];
}

function getHighlightColor(note: Note): string {
  if (isServerSide()) {
    return DEFAULT_COLORS[note.access] || DEFAULT_FALLBACK_COLOR;
  }

  switch (note.access) {
    case "public":
      return computeCssVar("--note-public", DEFAULT_COLORS.public);
    case "private":
      return computeCssVar("--note-private", DEFAULT_COLORS.private);
    case "organization":
      return computeCssVar("--note-org", DEFAULT_COLORS.organization);
    default:
      return computeCssVar("--gray-2", DEFAULT_FALLBACK_COLOR);
  }
}

export function loadImage(src: string): Promise<HTMLImageElement> {
  return new Promise((resolve, reject) => {
    const image = new Image();
    image.addEventListener("load", () => resolve(image));
    image.addEventListener("error", reject);
    image.src = src;
  });
}

interface NoteCoordinates {
  x: number;
  y: number;
  width: number;
  height: number;
}

export function transformNoteCoordinates(
  note: Note,
  containerWidth: number,
  containerHeight: number,
  scale: number = 1,
): NoteCoordinates {
  return {
    x: note.x1 * containerWidth * scale,
    y: note.y1 * containerHeight * scale,
    width: width(note) * containerWidth * scale,
    height: height(note) * containerHeight * scale,
  };
}

export async function renderImage(
  note: Note,
  canvas: HTMLCanvasElement,
  document: Document,
) {
  const page_number = note.page_number;
  const context = getCanvasContext(canvas);
  if (!context) {
    return;
  }

  const src = pageImageUrl(document, page_number, SIZE);
  const image = await loadImage(src.href);

  // Calculate dimensions for full-width context with padding
  const dimensions = calculateImageContextDimensions(note, image);

  // Set canvas to full width with padded height
  setupCanvas(canvas, image.width, dimensions.sourceHeight);

  // Draw the full-width context
  context.drawImage(
    image,
    0, // source x - full width
    dimensions.sourceY, // source y - with padding
    image.width, // source width - full width
    dimensions.sourceHeight, // source height - note height + padding
    0, // dest x
    0, // dest y
    image.width, // dest width
    dimensions.sourceHeight, // dest height
  );

  // Draw highlight rectangle on canvas
  // The highlight Y position is the note's Y position minus the source Y offset
  const highlightY = note.y1 * image.height - dimensions.sourceY;
  drawHighlight(
    context,
    getHighlightColor(note),
    note.x1 * image.width,
    highlightY,
    dimensions.noteWidth,
    dimensions.noteHeight,
  );
}

export async function renderPDF(
  note: Note,
  scale: number,
  canvas: HTMLCanvasElement,
  pdf: PDFDocumentProxy,
) {
  const page_number = note.page_number + 1; // note pages are 0-indexed
  const context = getCanvasContext(canvas);
  if (!context) {
    return;
  }

  const page = await pdf.getPage(page_number);
  const [, , w, h] = page.view;

  if (!w || !h) {
    console.error("Missing page dimensions when rendering PDF in note.");
    return;
  }

  // Calculate dimensions for context with padding
  const dimensions = calculatePDFContextDimensions(note, w, h, scale);

  const viewport = page.getViewport({
    scale,
    offsetX: 0, // full width
    offsetY: -dimensions.sourceY,
  });

  const dpr = window?.devicePixelRatio ?? 1;
  const transform = dpr !== 1 ? [dpr, 0, 0, dpr, 0, 0] : undefined;

  // Set the pixel dimensions of the canvas to full width with padding
  setupCanvas(canvas, w * scale, dimensions.sourceHeight, dpr);

  const renderTask = page.render({
    canvasContext: context,
    viewport,
    transform,
  });

  await renderTask.promise;

  // Draw highlight rectangle on canvas after PDF rendering is complete
  // The highlight Y position is the note's Y position minus the context offset
  const highlightY = note.y1 * h * scale - dimensions.sourceY;
  drawHighlight(
    context,
    getHighlightColor(note),
    note.x1 * w * scale * dpr,
    highlightY * dpr,
    dimensions.noteWidth * dpr,
    dimensions.noteHeight * dpr,
  );
}
