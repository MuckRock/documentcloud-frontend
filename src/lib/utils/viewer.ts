import type { Document, Note, ViewerMode } from "$lib/api/types";
import { canonicalUrl, pageHashUrl } from "../api/documents";
import { noteHashUrl } from "../api/notes";

interface ViewerHrefOptions {
  document?: Document;
  page?: number;
  note?: Note;
  mode?: ViewerMode;
  embed?: boolean;
}

export function getViewerHref(options: ViewerHrefOptions = {}) {
  const { document, page, note, mode = "document", embed = false } = options;
  let hash = "";
  if (page) hash = pageHashUrl(page);
  if (note) hash = noteHashUrl(note);
  if (document) {
    // If we have the document, we can provide an absolute URL
    let url = canonicalUrl(document);
    url.searchParams.set("mode", mode);
    if (embed) url.searchParams.set("embed", "1");
    if (hash) url.hash = hash;
    return url.href;
  } else {
    // If we don't, we can provide a relative URL
    let href = `?mode=${mode}`;
    if (embed) href += "&embed=1";
    if (hash) href += hash;
    return href;
  }
}

/**
 * Return a numeric scale based on intrinsic page size and container size
 * @param width Original document width
 * @param height Original document height
 * @param container
 * @param scale
 */
export function fitPage(
  width: number,
  height: number,
  container: HTMLElement,
  scale: number | "width" | "height",
): number {
  if (typeof scale === "number") return scale;
  if (!container) return 1;

  // const [x1, y1, width, height] = page.view;
  const { clientWidth, clientHeight } = container;

  return scale === "width" ? clientWidth / width : clientHeight / height;
}

/**
 * Parse page_spec into width and height of each page
 *
 * @param pageSpec A string encoding page dimensions in a compact format
 * @returns an array of [width, height] tuples
 */
export function pageSizes(pageSpec: string): [width: number, height: number][] {
  // Handle empty page spec
  if (pageSpec.trim().length == 0) return [];

  const parts = pageSpec.split(";");
  return parts.reduce((sizes, part, i) => {
    const [size, range] = part?.split(":");
    const [width, height] = size?.split("x").map(parseFloat);

    range?.split(",").forEach((rangePart) => {
      if (rangePart.includes("-")) {
        const [start, end] = rangePart.split("-").map((x) => parseInt(x, 10));
        for (let page = start; page <= end; page++) {
          sizes[page] = [width, height];
        }
      } else {
        const page = parseInt(rangePart, 10);
        sizes[page] = [width, height];
      }
    });

    return sizes;
  }, Array(parts.length));
}
