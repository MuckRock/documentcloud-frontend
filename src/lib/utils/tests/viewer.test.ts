import { describe, it, expect, vi } from "vitest";
import { document } from "@/test/fixtures/documents";
import { note } from "@/test/fixtures/notes";
import { getViewerHref } from "../viewer";
import {
  canonicalUrl,
  pageHashUrl,
  READING_MODES,
  WRITING_MODES,
} from "@/lib/api/documents";
import { noteHashUrl } from "@/lib/api/notes";

describe("getViewerHref", () => {
  const docUrl = canonicalUrl(document);
  it("returns an absolute URL when provided a document", () => {
    expect(getViewerHref({ document }).startsWith(docUrl.href)).toBe(true);
  });
  it("returns a relative URL when document is missing", () => {
    expect(getViewerHref().startsWith("?")).toBe(true);
  });
  it("applies the mode as a query parameter", () => {
    const modes = [...READING_MODES, ...WRITING_MODES];
    modes.forEach((mode) => {
      expect(getViewerHref({ mode })).toEqual(`?mode=${mode}`);
      expect(getViewerHref({ document, mode })).toEqual(
        `${docUrl}?mode=${mode}`,
      );
    });
  });
  it("applies embed status query param if true", () => {
    expect(getViewerHref({ embed: true })).toMatch("embed=1");
    expect(getViewerHref({ embed: false })).not.toMatch("embed=0");
    // with document
    expect(getViewerHref({ document, embed: true })).toMatch("embed=1");
    expect(getViewerHref({ document, embed: false })).not.toMatch("embed=0");
  });
  it("adds a page to the URL", () => {
    expect(getViewerHref({ page: 10 })).toMatch(pageHashUrl(10));
    expect(getViewerHref({ document, page: 10 })).toMatch(pageHashUrl(10));
  });
  it("adds a note to the URL", () => {
    expect(getViewerHref({ note })).toMatch(noteHashUrl(note));
    expect(getViewerHref({ document, note })).toMatch(noteHashUrl(note));
  });
  it("overwrites the page hash with the note hash is both are provided", () => {
    expect(getViewerHref({ page: 10, note })).toMatch(noteHashUrl(note));
    expect(getViewerHref({ page: 10, note })).not.toMatch(pageHashUrl(10));
    expect(getViewerHref({ document, page: 10, note })).toMatch(
      noteHashUrl(note),
    );
    expect(getViewerHref({ document, page: 10, note })).not.toMatch(
      pageHashUrl(10),
    );
  });
});
