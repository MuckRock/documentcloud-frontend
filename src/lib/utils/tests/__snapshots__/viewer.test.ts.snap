// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zoom > getZoomLevels 1`] = `
[
  [
    "width",
    "zoom.fitWidth",
  ],
  [
    "height",
    "zoom.fitHeight",
  ],
  [
    0.5,
    "50%",
  ],
  [
    0.75,
    "75%",
  ],
  [
    1,
    "100%",
  ],
  [
    1.25,
    "125%",
  ],
  [
    1.5,
    "150%",
  ],
  [
    2,
    "200%",
  ],
]
`;

exports[`zoom > getZoomLevels 2`] = `
[
  [
    0.5,
    "50%",
  ],
  [
    0.75,
    "75%",
  ],
  [
    1,
    "100%",
  ],
  [
    1.25,
    "125%",
  ],
  [
    1.5,
    "150%",
  ],
  [
    2,
    "200%",
  ],
]
`;

exports[`zoom > getZoomLevels 3`] = `
[
  [
    "thumbnail",
    "zoom.thumbnail",
  ],
  [
    "small",
    "zoom.small",
  ],
  [
    "normal",
    "zoom.normal",
  ],
  [
    "large",
    "zoom.large",
  ],
]
`;

exports[`zoom > getZoomLevels 4`] = `[]`;

exports[`zoom > getZoomLevels 5`] = `
[
  [
    "width",
    "zoom.fitWidth",
  ],
  [
    "height",
    "zoom.fitHeight",
  ],
  [
    0.5,
    "50%",
  ],
  [
    0.75,
    "75%",
  ],
  [
    1,
    "100%",
  ],
  [
    1.25,
    "125%",
  ],
  [
    1.5,
    "150%",
  ],
  [
    2,
    "200%",
  ],
]
`;

exports[`zoom > getZoomLevels 6`] = `
[
  [
    "width",
    "zoom.fitWidth",
  ],
  [
    "height",
    "zoom.fitHeight",
  ],
  [
    0.5,
    "50%",
  ],
  [
    0.75,
    "75%",
  ],
  [
    1,
    "100%",
  ],
  [
    1.25,
    "125%",
  ],
  [
    1.5,
    "150%",
  ],
  [
    2,
    "200%",
  ],
]
`;
